src/crypto/OutgoingRoomKeyRequestManager.js

```diff
@@ -107,6 +110,79 @@ export default class OutgoingRoomKeyRequestManager {
         });
     }
 
+    /**
+     * Cancel room key requests, if any match the given details
+     *
+     * @param {module:crypto~RoomKeyRequestBody} requestBody
+     *
+     * @returns {Promise} resolves when the request has been updated in our
+     *    pending list.
+     */
+    cancelRoomKeyRequest(requestBody) {
+        return this._cryptoStore.getOutgoingRoomKeyRequest(
+            requestBody,
+        ).then((req) => {
+            if (!req) {
+                // no request was made for this key
+                return;
+            }
+
+            if (req.state === ROOM_KEY_REQUEST_STATES.UNSENT) {
+                // just delete it
+                console.log(
+                    'deleting unnecessary room key request for ' +
+                    stringifyRequestBody(requestBody),
+                );
+                return this._cryptoStore.deleteOutgoingRoomKeyRequest(
+                    req.requestId, ROOM_KEY_REQUEST_STATES.UNSENT,
+                );
+            }
+
+            if (req.state !== ROOM_KEY_REQUEST_STATES.SENT) {
+                // the only other option is cancellation_pending, in which case
+                // there is nothing else to do here
+            }
+
+            // send a cancellation.
+            return this._cryptoStore.updateOutgoingRoomKeyRequest(
+                req.requestId, ROOM_KEY_REQUEST_STATES.SENT, {
+                    state: ROOM_KEY_REQUEST_STATES.CANCELLATION_PENDING,
+                    cancellationTxnId: this._baseApis.makeTxnId(),
+                },
+            ).then((updatedReq) => {
+                if (!updatedReq) {
+                    // we raced with another tab to mark the request cancelled
+                    // - no point in sending another cancellation
```

Comment:
Aha, I see now. The bit I failed to understand before was:

> updatedReq being false means that the store didn't find the matching request in ROOM_KEY_REQUEST_STATES.SENT

coupled with:

> We know it was in that state

I guess my problem was that I failed to understand that:
```js
+            return this._cryptoStore.updateOutgoingRoomKeyRequest(
 +                req.requestId, ROOM_KEY_REQUEST_STATES.SENT, {
 +                    state: ROOM_KEY_REQUEST_STATES.CANCELLATION_PENDING,
 +                    cancellationTxnId: this._baseApis.makeTxnId(),
 +                },
 +            )
```
means "transition to SENT -> CANCELLATION_PENDING and return a bool if you did so". Don't know how best to improve clarity here though.